presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionarySelectiveReader.java

```diff
@@ -89,14 +90,19 @@
 
     private byte[] stripeDictionaryData = EMPTY_DICTIONARY_DATA;
     private int[] stripeDictionaryOffsetVector = EMPTY_DICTIONARY_OFFSETS;
+    private byte[] rowGroupDictionaryData = EMPTY_DICTIONARY_DATA;
+    private int[] rowGroupDictionaryOffsetVector = EMPTY_DICTIONARY_OFFSETS;
     private byte[] currentDictionaryData = EMPTY_DICTIONARY_DATA;
+    private int[] currentDictionaryOffsetVector;
     private int[] stripeDictionaryLength = new int[0];
     private int[] rowGroupDictionaryLength = new int[0];
     private byte[] evaluationStatus;
+    private byte[] valueWithPadding;
 
     private int readOffset;
 
     private VariableWidthBlock dictionary = new VariableWidthBlock(1, wrappedBuffer(EMPTY_DICTIONARY_DATA), EMPTY_DICTIONARY_OFFSETS, Optional.of(new boolean[] {true}));
+    private boolean dictionaryWrapped;
```

Comment:
Not a huge fan of this dictionaryWrapped state variable. Makes this hard to read and we now need to account manually when we create the dictionary and it is not wrapped. This makes it prone to future bugs if devs don't know what that this var is required to be updated. We already have a flag for stripeDictionaryOpen and this adds extra states. Let's put this beside the currentDictionarySize and add a comment for when to update this variable?

Reply:
@sachdevs I added the following comments:
```
 private boolean stripeDictionaryOpen;
    // The dictionaries will be wrapped in getBlock(). It's set to false when opening a new dictionary (be it stripe dictionary or rowgroup dictionary). When there is only stripe
    // dictionary but no rowgroup dictionaries, we shall set it to false only when opening the stripe dictionary while not for every rowgroup. It is set to true when the dictionary
    // is wrapped up in wrapDictionaryIfNecessary().
    private boolean dictionaryWrapped;
```

> Makes this hard to read and we now need to account manually when we create the dictionary and it is not wrapped. 

Correction: It's not " when we create the dictionary". The dictionary was not created but just read from the input streams in openRowGroup().

This variable is required because we don't want to do memory allocation for every rowgroup unless absolutely required(ie. if no rows are selected for the current block then no need to create the dictionary). Furthermore, it needs to be set false explicitly because we don't want to allocate memory for each row group when there is no rowgroup dictionary but only stripe dictionary. A stripe contains multiple rowgroups, and a rowgroup can produce multiple blocks. Dictionaries are only read once from the input streams, but we don't want to copy the read data into a real dictionary for every block/rowgroup. There has to be a state telling us if the DictionaryBlock has been constructed or not for the current output block. The idea is like a state flag: when a new dictionary data is read, set to false; At the first time of DictionaryBlock creation, set it to true. Do you have better idea to not use the state variable without refactoring the whole class too much?

Reply:
Thanks for the detailed explanation.
> This variable is required because we don't want to do memory allocation for every rowgroup unless absolutely required
Yeah I agree that it is necessary. Without refactoring it's hard to get rid of this (I would refactor this into two separate classes probably, no other way to get rid of the var) - which is why I think adding a comment is sufficient for now.